compiler/rustc_lint/src/impl_trait_overcaptures.rs

```diff
@@ -221,25 +279,60 @@ impl<'tcx> TypeVisitor<TyCtxt<'tcx>> for VisitOpaqueTypes<'tcx> {
                     if variances[param.index as usize] != ty::Invariant {
                         continue;
                     }
+
+                    let arg = opaque_ty.args[param.index as usize];
                     // We need to turn all `ty::Param`/`ConstKind::Param` and
                     // `ReEarlyParam`/`ReBound` into def ids.
-                    captured.insert(extract_def_id_from_arg(
-                        self.tcx,
-                        generics,
-                        opaque_ty.args[param.index as usize],
-                    ));
+                    captured.insert(extract_def_id_from_arg(self.tcx, generics, arg));
+
+                    captured_regions.extend(arg.as_region());
                 }
                 current_def_id = generics.parent;
             }
 
             // Compute the set of in scope params that are not captured. Get their spans,
             // since that's all we really care about them for emitting the diagnostic.
-            let uncaptured_spans: Vec<_> = self
+            let mut uncaptured_args: FxIndexSet<_> = self
                 .in_scope_parameters
                 .iter()
-                .filter(|def_id| !captured.contains(*def_id))
-                .map(|def_id| self.tcx.def_span(def_id))
+                .filter(|&(def_id, _)| !captured.contains(def_id))
+                .collect();
+
+            // These are args that we know are likely fine to "overcapture", since they can be
+            // contravariantly shortened to one of the already-captured lifetimes that they
+            // outlive.
+            let covariant_long_args: FxIndexSet<_> = uncaptured_args
+                .iter()
+                .copied()
+                .filter(|&(def_id, kind)| {
+                    let Some(ty::Bivariant | ty::Contravariant) = self.variances.get(def_id) else {
+                        return false;
+                    };
+                    let DefKind::LifetimeParam = self.tcx.def_kind(def_id) else {
```

Comment:
is this reachable?

Reply:
No, not due to how we construct the variances I guess.